docs/designs/concepts.md

```diff
@@ -0,0 +1,180 @@
+# Concepts Design Doc
+
+This document describes core concepts of ElasticDL.
+
+An ElasticDL model consists of two kinds of parameters:
+
+1. dense parameters, and
+1. embedding tables.
+
+A dense parameter is a dense tensor with a name.  An embedding table is a map from some ID to embedding vectors.  An embedding table also has a name.  Formalizing the concepts, we have
+
+- model = {dense parameter} + {embedding table}
+- dense parameter = tensor + name
+- embedding table = {id, tensor} + name
+
+where the curly braces denote zero or more.
+
+It is notable that in practice, we use the rule
+
+- embedding table = tensor + index + name
+- index = map[ID]address
+
+because we append all embedding vectors into the tensor and use index, which is a map from ID to the starting address of the embedding vector in the tensor.
+
+To update the model, workers compute and report gradients.  Accordingly, we have two kinds of gradients:
+
+1. dense gradient, and
+1. embedding table gradient
+
+The content of dense gradient is the same as that of the dense parameter.  The content of embedding table gradient is the same as that of the embedding table.
+
+Please be aware that if some embedding table maps zero-based successive ID values to embedding vectors.  Such kind of embedding table, if not too large, could be represented by a dense parameters.  In the rest of the design, the concept embedding table refers to a sparse/general embedding table, where the ID might not be zero-based successive values, and even not an numerical value.
+
+Let's make a short summary, following is all the core concepts of ElasticDL include:
+
+- model = {dense parameter} + {embedding table}
+- dense parameter = tensor + name
+- embedding table = tensor + index + name
+- index = map[ID]address
+
+## In-Memory Representation
+
+EmbeddingTable is initialized lazily in PS. Worker sends item ids to request embedding vectors from PS. If the ID is not found in PS, the corresponding embedding vector will be initialized and sent back to the worker. In order to query and insert embedding vectors efficiently, we use a map data structure to represent EmbeddingTable.
+
+```go
```

Comment:
I would suggest that we use protobuf instead of Go code in the design doc because protobuf is designed to be an IDL.

Reply:
Here, I want to clarify the reason why we still need some extra in-memory data structure in Go, especially for embedding table. There are two reasons:

1.  Embedding table is usually a large dense tensor. In forward pass of embedding layer, worker will not pull the full table, but sends an array of item IDs to PS. Then the PS looks up the embeddig table, and sends corresponding embedding vectors back to worker. The worker will get a subset of the whole table. In the backward pass of embeddig layer, the gradients are also associated with IDs. Only IDs which are used in forward pass make contribution to the gradients.

TensorFlow introduce a data structure called [IndexedSlices](https://www.tensorflow.org/api_docs/python/tf/IndexedSlices) to represent the subset of a big dense tensor.
 The IndexedSlices has two filed, Tensor and Indices. Indices is a 1-dim array.

2.  In our design, the embedding table is initialized lazily in PS. And the size of embedding table is also unknown. So we could not create a big dense tensor simplely. When looking IDs from workers, if the ID is not found in the table, the corresponding embedding vector will be created and initialized.  In order to query and insert embedding vectors efficiently, we use a map data structure to represent EmbeddingTable.

In PS side, the embedding table parameter is a map data structure. In worker side, we pull the subset of the embedding table from PS, and push the gradients associated with IDs to PS.

We define a `IndexedSlices` message for the embedding table related RPC between worker and PS.

Following is the design:

```proto
message Tensor {
  repeated int64 dims = 1;
  bytes content = 2;
  TensorElementType dtype = 3;
}

message IndexedSlices {
  Tensor tensor = 1;
  repeated int64 indices = 2;
}

message Model {
  int32 version = 1; // model updated times
  map<string, Tensor> dense_parameters = 2;
  map<string, IndexedSlices> indexed_slices = 3;
}
```

For PS side, we propose following in-memory data structure in Go:

```go
import "elasticdl.org/elasticdl/pkg/proto"

type EmbeddingTable struct {
	Name          string
	Dim             int64
	Initializer     string
	EmbeddingVector map[int64]*proto.Tensor
}

type Model struct {
  Version           int32
  InitStatus        bool
  DenseParameters   map[string]*proto.Tensor
  EmbeddingTables   map[string]*EmbeddingTable
}
```

Reply:
After the phone call, I got the point.

On the parameter server, we'd prefer to maintain each embedding table as a map from ID to embedding vectors.  With such a data structure, it is efficient to allocate memory for new embedding vectors.  On the contrary, we'd concatenate embedding vectors into protobuf messages for parameter pulling and gradient pushing.  We cannot use concatenated embedding vectors as the in-memory data structure on the PS, because allocating new embedding vectors involves resize the space of concatenated embedding vectors.

Reply:
```protobuf
message IndexedSlices {
  Tensor concat_embedding_vecs = 1;
  repeated uint64 ids = 2;
}
```